src/js_macro_folder.rs

```diff
@@ -25,7 +25,7 @@ impl<'a> JsMacroFolder<'a> {
     }
 
     fn create_message_descriptor_from_tokens(&mut self, tokens: Vec<MsgToken>) -> Expr {
-      let parsed = MessageBuilder::parse(tokens);
+      let parsed = MessageBuilder::parse(tokens, &self.ctx);
```

Comment:
I don'y think it's a good idea to pass the Folder / Visitor object into the MessageBuilder, they are completely decoupled. I think reading of the placeholders should happend on somewhere where tokens are calculated.  MessageBuilder shouldn't know anything about that.

Reply:
The best place to parse explicit placeholders are: 

- For JSX visitor in `src/jsx_visitor.rs` `visit_jsx_expr_container`
- For js visitor in `src/macro_utils.rs:132` `tokenize_tpl`

Might be some new util to try parse an expression as labeled placeholder could be added to `macro_utils`

Reply:
Hello Timofei, sorry for a slight delay :D
Added another variant for MsgToken and moved parsing logic as you requested. Looks much better now.